src/Sentry/Internal/UnsampledTransaction.cs

```diff
@@ -0,0 +1,103 @@
+using Sentry.Extensibility;
+
+namespace Sentry.Internal;
+
+/// <summary>
+/// We know already, when starting a transaction, whether it's going to be sampled or not. When it's not sampled, we can
+/// avoid lots of unecessary processing. The only thing we need to track is the number of spans that would have been
+/// created (the client reports detailing discarded events includes this detail).
+/// </summary>
+internal sealed class UnsampledTransaction : NoOpTransaction
```

Comment:
I was hoping we'd get away without having to create an instance for each transaction but I also dunno how we'd deal with trace propagation in that case. And since most of the allocation are on the spans anyway this prob going to shave off enough GC pressure